feat: add decimal validation for numeric precision and scale supported by Spanner

vi3k6i5 opened the pull request:
feat: add decimal validation for numeric precision and scale supported by spanner
Fixes #339 🦕

olavloite commented on google/cloud/spanner_dbapi/parse_utils.py:
```diff
@@ -509,11 +518,37 @@ def sql_pyformat_args_to_spanner(sql, params):
     return sanitize_literals_for_upload(sql), named_args
 
 
+def assert_numeric_precision_and_scale(value):
```
I don't know whether this is normal in the Python client, but it seems a little strange that this function is called for __all__ parameter types, and not only numeric values. I would have expected either:
1. That this function is only called for numeric values (i.e. the `if isinstance(value, decimal.Decimal)` would be outside of the function).
2. Or: That there is a generic 'assert_valid_value(value)` function that first checks the type of the input parameter, and then calls the appropriate validation function for that type.

vi3k6i5 replied:
makes sense added an instance check before the method is called.

olavloite reviewed (approved):
This seems good to my untrained Python eyes, with a small question on whether it is logical to call the validate function for all values, or whether it would be better to move the if-check for the type out of the validation function.
(You should in any case await approval from other reviewers before merging)

larkee commented on google/cloud/spanner_dbapi/parse_utils.py:
```diff
@@ -509,11 +518,37 @@ def sql_pyformat_args_to_spanner(sql, params):
     return sanitize_literals_for_upload(sql), named_args
 
 
+def assert_numeric_precision_and_scale(value):
```
I think this logic would make more sense to be added to the base client [here](https://github.com/googleapis/python-spanner/blob/master/google/cloud/spanner_v1/_helpers.py#L131-L132).

vi3k6i5 replied:
moved it to base client.

vi3k6i5 commented:
> This seems good to my untrained Python eyes, with a small question on whether it is logical to call the validate function for all values, or whether it would be better to move the if-check for the type out of the validation function.
> (You should in any case await approval from other reviewers before merging)

moved the method after a check. and also moved it to base client as @larkee suggested.

larkee commented on google/cloud/spanner_v1/_helpers.py:
```diff
@@ -87,6 +97,27 @@ def _merge_query_options(base, merge):
+    Read more here: https://cloud.google.com/spanner/docs/data-types#numeric_type
+
+    Asserts that input numeric field is within spanner supported range.
+    :type value: Any
```
```suggestion
    :type value: decimal.Decimal
```

vi3k6i5 replied:
fixed.

larkee commented on google/cloud/spanner_v1/_helpers.py:
```diff
@@ -87,6 +97,27 @@ def _merge_query_options(base, merge):
+    This number can be optionally prefixed with a plus or minus sign.
+    Read more here: https://cloud.google.com/spanner/docs/data-types#numeric_type
+
+    Asserts that input numeric field is within spanner supported range.
```
```suggestion
    Asserts that input numeric field is within Spanner supported range.
```
Also this should be moved to the top of the docstring. That way you have an explanation of what function does, followed by why it's needed.

vi3k6i5 replied:
Done.

larkee commented on google/cloud/spanner_v1/_helpers.py:
```diff
@@ -87,6 +97,27 @@ def _merge_query_options(base, merge):
+    :type value: Any
+    :param value: The value to check for Cloud Spanner compatibility.
+
+    :raises NotSupportedError: if value is not within supporteed precision or scale of spanner.
```
```suggestion
    :raises NotSupportedError: if value is not within supported precision or scale of Spanner.
```

vi3k6i5 replied:
fixed comment.

larkee commented on tests/unit/test__helpers.py:
```diff
@@ -233,6 +233,66 @@ def test_w_unknown_type(self):
+        ]
+        for value in cases:
+            with self.subTest(value=value):
+                value = decimal.Decimal("9999999999999999999999999999.999999999")
```
I don't think this line is needed. The value should be coming from the cases.

vi3k6i5 replied:
changed to read from cases.

larkee reviewed (approved):
LGTM :+1:  You could remove `test_w_numeric` if you wanted since `test_w_numeric_precision_and_scale_valid` should have more coverage.
